src/media-segment-request.js

```diff
@@ -0,0 +1,281 @@
+      code: REQUEST_ERRORS.FAILURE,
+      xhr: request
+    }, segment);
+  }
+
+  const view = new DataView(response);
+
+  segment.key.bytes = new Uint32Array([
+    view.getUint32(0),
+    view.getUint32(4),
+    view.getUint32(8),
+    view.getUint32(12)
+  ]);
+  return finishProcessingFn(null, segment);
+};
+
+const handleInitSegmentResponse = (segment, finishProcessingFn) => (error, request) => {
+  const errorObj = handleErrors(error, request);
+
+  if (errorObj) {
+    return finishProcessingFn(errorObj, segment);
+  }
+
+  segment.map.bytes = new Uint8Array(request.response);
+
+  if (!Array.isArray(segment.bandwidth)) {
+    segment.bandwidth = [request.bandwidth];
+  } else {
+    segment.bandwidth.push(request.bandwidth);
+  }
+
+  return finishProcessingFn(null, segment);
+};
+
+const handleSegmentResponse = (segment, finishProcessingFn) => (error, request) => {
+  const errorObj = handleErrors(error, request);
+
+  if (errorObj) {
+    return finishProcessingFn(errorObj, segment);
+  }
+  segment.stats = getRequestStats(request);
+
+  if (segment.key) {
+    segment.encryptedBytes = new Uint8Array(request.response);
+  } else {
+    segment.bytes = new Uint8Array(request.response);
+  }
+
+  return finishProcessingFn(null, segment);
+};
+
+const maybeDecrypt = (decrypter, segment, callback) => {
+  // Keep track of when all of the requests have completed
+  segment.endOfAllRequests = Date.now();
+
+  // If the segment is not encrypted, just continue
+  if (!segment.encryptedBytes) {
+    return callback(null, segment);
+  }
+
+  const decryptionId = 'segment-request-' + Math.random().toString(36);
+  const decryptionHandler = (event) => {
+    if (event.data.source === decryptionId) {
+      decrypter.removeEventListener('message', decryptionHandler);
+      const decrypted = event.data.decrypted;
+
+      segment.bytes = new Uint8Array(decrypted.bytes,
+                                     decrypted.byteOffset,
+                                     decrypted.byteLength);
+      return callback(null, segment);
+    }
+  };
+
+  decrypter.addEventListener('message', decryptionHandler);
+
+  // this is an encrypted segment
+  // incrementally decrypt the segment
+  decrypter.postMessage(createTransferableMessage({
+    source: decryptionId,
+    encrypted: segment.encryptedBytes,
+    key: segment.key.bytes,
+    iv: segment.key.iv
+  }), [
+    segment.encryptedBytes.buffer,
+    segment.key.bytes.buffer
+  ]);
+};
+
+const waitForCompletion = (activeXhrs, decrypter, callback) => {
+  let errors = [];
+  let count = 0;
+
+  return (error, segment) => {
+    // errors have to be unshifted to make sure the original error - the one
+    // that resulted in several aborts - ends up at the start of the array for
+    // ease of use downstream
+    if (error) {
+      // If there are errors, we have to abort any outstanding requests
+      abortAll(activeXhrs);
+      errors.unshift(error);
```

Comment:
You might be right. It looks like I was seeing sinon's mock xhr results (which are synchronous) and was confusing it with actual behavior.